File: conformance/tests/p384-wallet/p384-wallet.runar.rb
require 'runar'

class P384Wallet < Runar::SmartContract
  prop :ecdsa_pub_key_hash, Addr
  prop :p384_pub_key_hash, ByteString

  def initialize(ecdsa_pub_key_hash, p384_pub_key_hash)
    super(ecdsa_pub_key_hash, p384_pub_key_hash)
    @ecdsa_pub_key_hash = ecdsa_pub_key_hash
    @p384_pub_key_hash = p384_pub_key_hash
  end

  runar_public p384_sig: ByteString, p384_pub_key: ByteString, sig: Sig, pub_key: PubKey
  def spend(p384_sig, p384_pub_key, sig, pub_key)
    assert hash160(pub_key) == @ecdsa_pub_key_hash
    assert check_sig(sig, pub_key)
    assert hash160(p384_pub_key) == @p384_pub_key_hash
    assert verify_ecdsa_p384(sig, p384_sig, p384_pub_key)
  end
end

File: examples/zig/p384-wallet/P384Wallet.runar.zig
const runar = @import("runar");

pub const P384Wallet = struct {
    pub const Contract = runar.SmartContract;

    ecdsaPubKeyHash: runar.Addr,
    p384PubKeyHash: runar.ByteString,

    pub fn init(ecdsaPubKeyHash: runar.Addr, p384PubKeyHash: runar.ByteString) P384Wallet {
        return .{
            .ecdsaPubKeyHash = ecdsaPubKeyHash,
            .p384PubKeyHash = p384PubKeyHash,
        };
    }

    pub fn spend(
        self: *const P384Wallet,
        p384Sig: runar.ByteString,
        p384PubKey: runar.ByteString,
        sig: runar.Sig,
        pubKey: runar.PubKey,
    ) void {
        runar.assert(runar.bytesEq(runar.hash160(pubKey), self.ecdsaPubKeyHash));
        runar.assert(runar.checkSig(sig, pubKey));
        runar.assert(runar.bytesEq(runar.hash160(p384PubKey), self.p384PubKeyHash));
        runar.assert(runar.verifyECDSA_P384(sig, p384Sig, p384PubKey));
    }
};
